File: typescript/strain/strain.ts
function keep<T>(xs: T[], predicate: (e: T) => boolean): T[] {
  return xs.filter((x) => predicate(x))
}

function discard<T>(xs: T[], predicate: (e: T) => boolean): T[] {
  return xs.filter((x) => !predicate(x))
}

export { keep, discard }

File: ruby/strain/strain.rb
class Array
  def keep(&block)
    each_with_object([]) do |elt, result|
      result << elt if block.call(elt)
    end
  end

  def discard(&block)
    each_with_object([]) do |elt, result|
      result << elt unless block.call(elt)
    end
  end
end
